File: circuito_rlc_paralelo.py
#!/usr/bin/env python3
"""
ANÁLISE DO TRANSIENTE EM CIRCUITOS RLC PARALELO
Baseado no Capítulo 8 - Fundamentos de Circuitos Elétricos (Sadiku)
Autor: Python para Engenharia Elétrica
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp
import sympy as sp
from sympy import symbols, Function, Eq, dsolve, diff, simplify, integrate
import pandas as pd
from dataclasses import dataclass
from typing import Tuple, Dict, Any

@dataclass
class CircuitoRLC_Paralelo:
    """Classe para análise de circuitos RLC paralelo"""
    
    def __init__(self, R: float, L: float, C: float, I: float = 0):
        """
        Inicializa o circuito RLC paralelo
        
        Parâmetros:
        -----------
        R : float
            Resistência (Ω)
        L : float
            Indutância (H)
        C : float
            Capacitância (F)
        I : float, opcional
            Corrente da fonte DC (A)
        """
        self.R = R
        self.L = L
        self.C = C
        self.I = I
        
        # Parâmetros característicos
        self.alpha = 1 / (2 * R * C)  # Coeficiente de amortecimento (paralelo)
        self.omega0 = 1 / np.sqrt(L * C)  # Frequência natural (rad/s)
        self.zeta = self.alpha / self.omega0  # Fator de amortecimento
        
        # Determinar tipo de resposta
        self.tipo_resposta = self._classificar_resposta()
        
    def _classificar_resposta(self) -> str:
        """Classifica o tipo de resposta do circuito"""
        if self.zeta > 1:
            return "superamortecido"
        elif abs(self.zeta - 1) < 1e-9:
            return "criticamente amortecido"
        else:
            return "subamortecido"
    
    def calcular_parametros(self) -> Dict[str, float]:
        """Calcula e retorna os parâmetros característicos"""
        if self.tipo_resposta == "subamortecido":
            omega_d = np.sqrt(self.omega0**2 - self.alpha**2)
            Q = self.R * np.sqrt(self.C / self.L)  # Fator de qualidade (paralelo)
            t_settle_2 = 4 / self.alpha
            t_settle_5 = 3 / self.alpha
        else:
            omega_d = 0
            Q = 0
            t_settle_2 = 4 / self.alpha
            t_settle_5 = 3 / self.alpha
            
        return {
            'alpha': self.alpha,
            'omega0': self.omega0,
            'zeta': self.zeta,
            'omega_d': omega_d if self.tipo_resposta == "subamortecido" else 0,
            'Q': Q,
            't_settle_2%': t_settle_2,
            't_settle_5%': t_settle_5,
            'tipo': self.tipo_resposta
        }
    
    def resolver_sistema_edos(self, t_span: Tuple[float, float], 
                             cond_iniciais: Tuple[float, float],
                             n_points: int = 1000) -> Dict[str, np.ndarray]:
        """
        Resolve o sistema de EDOs numericamente
        
        Parâmetros:
        -----------
        t_span : tuple
            Intervalo de tempo (t_inicial, t_final)
        cond_iniciais : tuple
            Condições iniciais (v0, iL0)
        n_points : int
            Número de pontos na solução
            
        Retorna:
        --------
        dict : Dicionário com as soluções
        """
        # Sistema de EDOs: dv/dt = (I - v/R - iL)/C, diL/dt = v/L
        def sistema_edos(t, y, I, R, L, C):
            v, iL = y
            dv_dt = (I - v/R - iL) / C
            diL_dt = v / L
            return [dv_dt, diL_dt]
        
        # Vetor de tempo
        t = np.linspace(t_span[0], t_span[1], n_points)
        
        # Resolver usando solve_ivp
        sol = solve_ivp(
            lambda t, y: sistema_edos(t, y, self.I, self.R, self.L, self.C),
            t_span,
            cond_iniciais,
            t_eval=t,
            method='RK45',
            rtol=1e-8,
            atol=1e-10
        )
        
        # Extrair resultados
        t_sol = sol.t
        v_sol = sol.y[0]
        iL_sol = sol.y[1]
        
        # Calcular outras grandezas
        iR_sol = v_sol / self.R
        iC_sol = self.I - iR_sol - iL_sol
        pR_sol = v_sol**2 / self.R
        
        # Energias
        W_C = 0.5 * self.C * v_sol**2
        W_L = 0.5 * self.L * iL_sol**2
        W_total = W_C + W_L
        
        return {
            't': t_sol,
            'v': v_sol,
            'iL': iL_sol,
            'iR': iR_sol,
            'iC': iC_sol,
            'pR': pR_sol,
            'W_C': W_C,
            'W_L': W_L,
            'W_total': W_total
        }
    
    def plotar_resposta(self, solucao: Dict[str, np.ndarray], 
                       titulo: str = "Resposta Transitória - Circuito RLC Paralelo"):
        """
        Plota os resultados da simulação
        
        Parâmetros:
        -----------
        solucao : dict
            Dicionário com as soluções do sistema
        titulo : str
            Título do gráfico
        """
        fig, axes = plt.subplots(3, 3, figsize=(15, 12))
        
        # 1. Tensão no circuito
        ax = axes[0, 0]
        ax.plot(solucao['t'], solucao['v'], 'b-', linewidth=2)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('Tensão v(t) (V)')
        ax.set_title('Tensão no Circuito')
        ax.grid(True, alpha=0.3)
        ax.axhline(y=0, color='k', linestyle='-', linewidth=0.5)
        
        # 2. Correntes nos componentes
        ax = axes[0, 1]
        ax.plot(solucao['t'], solucao['iR'], 'r-', label='i_R', linewidth=2)
        ax.plot(solucao['t'], solucao['iL'], 'g-', label='i_L', linewidth=2)
        ax.plot(solucao['t'], solucao['iC'], 'b-', label='i_C', linewidth=2)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('Corrente (A)')
        ax.set_title('Correntes nos Componentes')
        ax.legend()
        ax.grid(True, alpha=0.3)
        
        # 3. Potência no resistor
        ax = axes[0, 2]
        ax.plot(solucao['t'], solucao['pR'], 'm-', linewidth=2)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('Potência (W)')
        ax.set_title('Potência Dissipada no Resistor')
        ax.grid(True, alpha=0.3)
        
        # 4. Energias armazenadas
        ax = axes[1, 0]
        ax.plot(solucao['t'], solucao['W_C'], 'b-', label='Capacitor', linewidth=2)
        ax.plot(solucao['t'], solucao['W_L'], 'g-', label='Indutor', linewidth=2)
        ax.plot(solucao['t'], solucao['W_total'], 'r-', label='Total', linewidth=2)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('Energia (J)')
        ax.set_title('Energia Armazenada')
        ax.legend()
        ax.grid(True, alpha=0.3)
        
        # 5. Diagrama de fases (v vs iL)
        ax = axes[1, 1]
        ax.plot(solucao['v'], solucao['iL'], 'b-', linewidth=1.5)
        ax.scatter(solucao['v'][0], solucao['iL'][0], color='r', s=100, 
                  zorder=5, label='Início')
        ax.scatter(solucao['v'][-1], solucao['iL'][-1], color='g', s=100,
                  zorder=5, label='Fim')
        ax.set_xlabel('Tensão v (V)')
        ax.set_ylabel('Corrente no indutor i_L (A)')
        ax.set_title('Diagrama de Fases (v vs i_L)')
        ax.legend()
        ax.grid(True, alpha=0.3)
        ax.axhline(y=0, color='k', linestyle='-', linewidth=0.5)
        ax.axvline(x=0, color='k', linestyle='-', linewidth=0.5)
        
        # 6. Tensão e derivada
        ax = axes[1, 2]
        ax.plot(solucao['t'], solucao['v'], 'b-', label='v(t)', linewidth=2)
        dv_dt = np.gradient(solucao['v'], solucao['t'])
        ax.plot(solucao['t'], dv_dt / np.max(np.abs(dv_dt)) * np.max(np.abs(solucao['v'])), 
                'r--', label='dv/dt (normalizado)', linewidth=1.5)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('Tensão (V)')
        ax.set_title('Tensão e Derivada')
        ax.legend()
        ax.grid(True, alpha=0.3)
        
        # 7. Comparação i_R, i_L, i_C em subplot
        ax = axes[2, 0]
        time_limited = solucao['t'] < min(0.1, np.max(solucao['t']))
        if np.any(time_limited):
            t_lim = solucao['t'][time_limited]
            ax.plot(t_lim, solucao['iR'][time_limited], 'r-', linewidth=2)
            ax.plot(t_lim, solucao['iL'][time_limited], 'g-', linewidth=2)
            ax.plot(t_lim, solucao['iC'][time_limited], 'b-', linewidth=2)
            ax.set_xlabel('Tempo (s)')
            ax.set_ylabel('Corrente (A)')
            ax.set_title('Correntes (zoom inicial)')
            ax.grid(True, alpha=0.3)
        
        # 8. Análise logarítmica do decaimento
        ax = axes[2, 1]
        envelope = np.exp(-self.alpha * solucao['t'])
        ax.semilogy(solucao['t'], np.abs(solucao['v']), 'b-', label='|v(t)|', linewidth=2)
        ax.semilogy(solucao['t'], envelope * np.max(np.abs(solucao['v'])), 
                   'r--', label='Envelope', linewidth=2)
        ax.set_xlabel('Tempo (s)')
        ax.set_ylabel('|Tensão| (V)')
        ax.set_title('Decaimento Exponencial')
        ax.legend()
        ax.grid(True, alpha=0.3, which='both')
        
        # 9. Informações do circuito
        ax = axes[2, 2]
        ax.axis('off')
        params = self.calcular_parametros()
        info_text = (
            f"R = {self.R:.2f} Ω\n"
            f"L = {self.L:.4f} H\n"
            f"C = {self.C:.4e} F\n"
            f"I = {self.I:.2f} A\n\n"
            f"α = {params['alpha']:.4f} Np/s\n"
            f"ω₀ = {params['omega0']:.4f} rad/s\n"
            f"ζ = {params['zeta']:.4f}\n"
            f"Tipo: {params['tipo']}\n"
        )
        if params['tipo'] == 'subamortecido':
            info_text += f"ω_d = {params['omega_d']:.4f} rad/s\n"
            info_text += f"Q = {params['Q']:.4f}\n"
            info_text += f"t_settle(2%) = {params['t_settle_2%']:.4f} s"
        
        ax.text(0.1, 0.5, info_text, transform=ax.transAxes,
                fontsize=11, verticalalignment='center',
                bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))
        
        plt.suptitle(titulo, fontsize=18, fontweight='bold')
        plt.tight_layout()
        plt.show()
        
        return fig


# Função principal para análise de circuito RLC paralelo
def analisar_rlc_paralelo():
    """Interface principal para análise de circuito RLC paralelo"""
    
    print("=" * 60)
    print("ANÁLISE DO TRANSIENTE - CIRCUITO RLC PARALELO")
    print("=" * 60)
    
    # Entrada de parâmetros
    print("\n--- PARÂMETROS DO CIRCUITO ---")
    
    # Valores padrão para teste rápido
    usar_exemplo = input("Usar exemplo rápido? (s/n): ").strip().lower()
    
    if usar_exemplo == 's':
        print("\nEscolha um exemplo:")
        print("1. Subamortecido (R=100Ω, L=0.1H, C=10μF, I=1A)")
        print("2. Criticamente amortecido (R=50Ω, L=0.1H, C=10μF, I=1A)")
        print("3. Superamortecido (R=10Ω, L=0.1H, C=10μF, I=1A)")
        
        escolha = input("Digite 1, 2 ou 3: ").strip()
        
        if escolha == '1':
            R, L, C, I = 100, 0.1, 10e-6, 1
            v0, iL0 = 0, 0
            t_final = 0.02
        elif escolha == '2':
            R, L, C, I = 50, 0.1, 10e-6, 1
            v0, iL0 = 0, 0
            t_final = 0.02
        elif escolha == '3':
            R, L, C, I = 10, 0.1, 10e-6, 1
            v0, iL0 = 0, 0
            t_final = 0.02
        else:
            print("Opção inválida. Usando exemplo subamortecido.")
            R, L, C, I = 100, 0.1, 10e-6, 1
            v0, iL0 = 0, 0
            t_final = 0.02
            
    else:
        try:
            R = float(input("Resistência R (Ω): "))
            L = float(input("Indutância L (H): "))
            C = float(input("Capacitância C (F): "))
            I = float(input("Corrente da fonte DC I (A): "))
            
            print("\n--- CONDIÇÕES INICIAIS ---")
            v0 = float(input("Tensão inicial no capacitor v(0) (V): "))
            iL0 = float(input("Corrente inicial no indutor iL(0) (A): "))
            
            print("\n--- TEMPO DE SIMULAÇÃO ---")
            t_final = float(input("Tempo final de simulação (s): "))
            
        except ValueError:
            print("Entrada inválida! Usando valores padrão.")
            R, L, C, I = 100, 0.1, 10e-6, 1
            v0, iL0 = 0, 0
            t_final = 0.02
    
    # Criar circuito
    circuito = CircuitoRLC_Paralelo(R, L, C, I)
    
    # Exibir parâmetros
    params = circuito.calcular_parametros()
    
    print("\n" + "=" * 60)
    print("PARÂMETROS CARACTERÍSTICOS")
    print("=" * 60)
    print(f"Resistência R = {R:.4f} Ω")
    print(f"Indutância L = {L:.4f} H")
    print(f"Capacitância C = {C:.4e} F")
    print(f"Corrente da fonte I = {I:.4f} A")
    print(f"\nCoeficiente de amortecimento α = {params['alpha']:.4f} Np/s")
    print(f"Frequência natural ω₀ = {params['omega0']:.4f} rad/s")
    print(f"Frequência natural f₀ = {params['omega0']/(2*np.pi):.4f} Hz")
    print(f"Fator de amortecimento ζ = α/ω₀ = {params['zeta']:.4f}")
    print(f"\nTIPO DE RESPOSTA: {params['tipo'].upper()}")
    
    if params['tipo'] == 'subamortecido':
        print(f"Frequência amortecida ω_d = {params['omega_d']:.4f} rad/s")
        print(f"Frequência amortecida f_d = {params['omega_d']/(2*np.pi):.4f} Hz")
        print(f"Fator de qualidade Q = {params['Q']:.4f}")
        print(f"Tempo de acomodação (2%) = {params['t_settle_2%']:.4f} s")
        print(f"Tempo de acomodação (5%) = {params['t_settle_5%']:.4f} s")
    
    # Resolver sistema
    print("\n" + "=" * 60)
    print("RESOLVENDO SISTEMA DE EDOs...")
    
    t_span = (0, t_final)
    cond_iniciais = (v0, iL0)
    
    solucao = circuito.resolver_sistema_edos(t_span, cond_iniciais, 1000)
    
    # Plotar resultados
    print("GERANDO GRÁFICOS...")
    circuito.plotar_resposta(solucao, f"Circuito RLC Paralelo - {params['tipo']}")
    
    print("\n" + "=" * 60)
    print("ANÁLISE CONCLUÍDA COM SUCESSO!")
    print("=" * 60)
    
    return circuito, solucao, params


if __name__ == "__main__":
    # Executar análise interativa
    circuito, solucao, params = analisar_rlc_paralelo()

File: circuito_rlc_paralelo.m
%% ============================================================
% ANÁLISE DO TRANSIENTE EM CIRCUITOS RLC PARALELO
% Baseado no Capítulo 8 - Fundamentos de Circuitos Elétricos (Sadiku)
% ============================================================

clear all; close all; clc;
fprintf('=== ANÁLISE DO TRANSIENTE - CIRCUITO RLC PARALELO ===\n\n');

%% 1. PARÂMETROS DO CIRCUITO
% ===========================
R = input('Resistência (Ω): ');
L = input('Indutância (H): ');
C = input('Capacitância (F): ');
I = input('Corrente da fonte (A): ');

% Condições iniciais
fprintf('\n--- Condições Iniciais ---\n');
v0 = input('Tensão inicial no capacitor v(0) (V): ');
iL0 = input('Corrente inicial no indutor i_L(0) (A): ');

% Tempo de simulação
t_final = input('\nTempo final de simulação (s): ');

%% 2. CÁLCULO DOS PARÂMETROS CARACTERÍSTICOS
% ===========================================
alpha = 1/(2*R*C);              % Coeficiente de amortecimento (paralelo)
omega0 = 1/sqrt(L*C);           % Frequência natural não amortecida
omega_d = sqrt(omega0^2 - alpha^2); % Frequência amortecida

% Classificação do circuito
fprintf('\n=== PARÂMETROS DO CIRCUITO ===\n');
fprintf('α = %.4f Np/s\n', alpha);
fprintf('ω₀ = %.4f rad/s\n', omega0);
fprintf('Frequência natural f₀ = %.4f Hz\n', omega0/(2*pi));

if alpha > omega0
    fprintf('\nSISTEMA SUPERAMORTECIDO (α > ω₀)\n');
    tipo = 1;
elseif abs(alpha - omega0) < 1e-6
    fprintf('\nSISTEMA CRITICAMENTE AMORTECIDO (α = ω₀)\n');
    tipo = 2;
else
    fprintf('\nSISTEMA SUBAMORTECIDO (α < ω₀)\n');
    fprintf('ω_d = %.4f rad/s\n', omega_d);
    fprintf('Frequência amortecida f_d = %.4f Hz\n', omega_d/(2*pi));
    tipo = 3;
end

% Fator de qualidade (paralelo)
if tipo == 3
    Q = R * sqrt(C/L);  % Fator de qualidade para paralelo
    t_settle_2 = 4/alpha;
    t_settle_5 = 3/alpha;
    fprintf('Fator de qualidade Q = %.4f\n', Q);
    fprintf('Tempo de acomodação (2%%) = %.4f s\n', t_settle_2);
    fprintf('Tempo de acomodação (5%%) = %.4f s\n', t_settle_5);
end

%% 3. SOLUÇÃO ANALÍTICA
% =====================
syms v(t)

% Equação diferencial: C*d²v/dt² + (1/R)*dv/dt + (1/L)*v = I'
% Para degrau de corrente: I' = 0 após t=0+
eqn = C*diff(v,t,2) + (1/R)*diff(v,t) + (1/L)*v == 0;

% Condições iniciais
cond1 = v(0) == v0;
% Para dv/dt(0): i_C(0) = C*dv/dt(0) = I - v0/R - iL0
dv0 = (I - v0/R - iL0)/C;
cond2 = diff(v) == dv0;

% Solução geral
if tipo == 1  % Superamortecido
    s1 = -alpha + sqrt(alpha^2 - omega0^2);
    s2 = -alpha - sqrt(alpha^2 - omega0^2);
    
    syms A1 A2
    v_sym = A1*exp(s1*t) + A2*exp(s2*t);
    
    eq1 = subs(v_sym, t, 0) == v0;
    eq2 = subs(diff(v_sym, t), t, 0) == dv0;
    
    [A1_sol, A2_sol] = solve([eq1, eq2], [A1, A2]);
    v_analitico = subs(v_sym, [A1, A2], [A1_sol, A2_sol]);
    
elseif tipo == 2  % Criticamente amortecido
    syms A1 A2
    v_sym = (A1 + A2*t)*exp(-alpha*t);
    
    eq1 = subs(v_sym, t, 0) == v0;
    eq2 = subs(diff(v_sym, t), t, 0) == dv0;
    
    [A1_sol, A2_sol] = solve([eq1, eq2], [A1, A2]);
    v_analitico = subs(v_sym, [A1, A2], [A1_sol, A2_sol]);
    
else  % Subamortecido
    syms B1 B2
    v_sym = exp(-alpha*t)*(B1*cos(omega_d*t) + B2*sin(omega_d*t));
    
    eq1 = subs(v_sym, t, 0) == v0;
    eq2 = subs(diff(v_sym, t), t, 0) == dv0;
    
    [B1_sol, B2_sol] = solve([eq1, eq2], [B1, B2]);
    v_analitico = subs(v_sym, [B1, B2], [B1_sol, B2_sol]);
end

v_analitico = simplify(v_analitico);
fprintf('\n=== SOLUÇÃO ANALÍTICA ===\n');
fprintf('v(t) = %s\n', char(v_analitico));

% Calcular correntes analiticamente
iR_analitico = v_analitico / R;
iC_analitico = C * diff(v_analitico, t);
iL_analitico = (1/L) * int(v_analitico, t) + iL0;

%% 4. SOLUÇÃO NUMÉRICA (ODE45)
% =============================
% Sistema: dv/dt = iC/C, diL/dt = v/L
% onde iC = I - v/R - iL

f = @(t, y) [
    (I - y(1)/R - y(2))/C;  % dv/dt
    y(1)/L                  % diL/dt
];

tspan = [0 t_final];
y0 = [v0; iL0];  % [v(0); iL(0)]

options = odeset('RelTol', 1e-8, 'AbsTol', 1e-10);
[t_num, y_num] = ode45(f, tspan, y0, options);

v_num = y_num(:, 1);
iL_num = y_num(:, 2);

% Calcular outras grandezas
iR_num = v_num / R;
iC_num = I - iR_num - iL_num;
pR_num = v_num.^2 / R;  % Potência dissipada

%% 5. VALIDAÇÃO
% =============
v_analitico_num = double(subs(v_analitico, t, t_num));
erro = max(abs(v_num - v_analitico_num));
fprintf('\n=== VALIDAÇÃO ===\n');
fprintf('Erro máximo entre soluções analítica e numérica: %.2e V\n', erro);

%% 6. VISUALIZAÇÃO DOS RESULTADOS
% ================================
figure('Position', [100, 100, 1200, 800]);

% Subplot 1: Tensão no circuito
subplot(3, 3, [1, 2]);
plot(t_num, v_num, 'b-', 'LineWidth', 2);
hold on;
plot(t_num, v_analitico_num, 'r--', 'LineWidth', 1.5);
xlabel('Tempo (s)', 'FontSize', 10);
ylabel('Tensão v(t) (V)', 'FontSize', 10);
title('Resposta Transitória - Tensão no Circuito Paralelo', 'FontSize', 12);
legend('Solução Numérica', 'Solução Analítica', 'Location', 'best');
grid on;

% Subplot 2: Correntes nos componentes
subplot(3, 3, [4, 5]);
plot(t_num, iR_num, 'r-', 'LineWidth', 2);
hold on;
plot(t_num, iL_num, 'g-', 'LineWidth', 2);
plot(t_num, iC_num, 'b-', 'LineWidth', 2);
xlabel('Tempo (s)', 'FontSize', 10);
ylabel('Corrente (A)', 'FontSize', 10);
title('Correntes nos Componentes', 'FontSize', 12);
legend('i_R(t)', 'i_L(t)', 'i_C(t)', 'Location', 'best');
grid on;

% Subplot 3: Potência dissipada
subplot(3, 3, [7, 8]);
plot(t_num, pR_num, 'm-', 'LineWidth', 2);
xlabel('Tempo (s)', 'FontSize', 10);
ylabel('Potência (W)', 'FontSize', 10);
title('Potência Dissipada no Resistor', 'FontSize', 12);
grid on;

% Subplot 4: Diagrama de fases
subplot(3, 3, [3, 6, 9]);
plot(v_num, iL_num, 'b-', 'LineWidth', 1.5);
hold on;
plot(v0, iL0, 'ro', 'MarkerSize', 10, 'MarkerFaceColor', 'r');
xlabel('Tensão v (V)', 'FontSize', 10);
ylabel('Corrente no indutor i_L (A)', 'FontSize', 10);
title('Diagrama de Fases (v vs i_L)', 'FontSize', 12);
grid on;

% Informações
annotation('textbox', [0.15, 0.85, 0.3, 0.1], 'String', ...
    sprintf('R = %.2f Ω\nL = %.4f H\nC = %.4f F\nI = %.2f A', R, L, C, I), ...
    'FitBoxToText', 'on', 'BackgroundColor', 'w');

if tipo == 1
    tipo_str = 'SUPERAMORTECIDO';
elseif tipo == 2
    tipo_str = 'CRITICAMENTE AMORTECIDO';
else
    tipo_str = 'SUBAMORTECIDO';
end

annotation('textbox', [0.65, 0.85, 0.3, 0.1], 'String', ...
    sprintf('TIPO: %s\nα = %.4f\nω₀ = %.4f\nQ = %.4f', ...
    tipo_str, alpha, omega0, Q), 'FitBoxToText', 'on', 'BackgroundColor', 'w');

%% 7. ANÁLISE DE ENERGIA
% ======================
W_C = 0.5 * C * v_num.^2;
W_L = 0.5 * L * iL_num.^2;
W_total = W_C + W_L;
W_R = cumtrapz(t_num, pR_num);

figure('Position', [100, 100, 1000, 600]);

subplot(2, 2, 1);
plot(t_num, W_C, 'b-', 'LineWidth', 2);
hold on;
plot(t_num, W_L, 'g-', 'LineWidth', 2);
plot(t_num, W_total, 'r-', 'LineWidth', 2);
xlabel('Tempo (s)');
ylabel('Energia (J)');
title('Energia Armazenada');
legend('W_C (capacitor)', 'W_L (indutor)', 'W_{total}', 'Location', 'best');
grid on;

subplot(2, 2, 2);
plot(t_num, W_R, 'm-', 'LineWidth', 2);
xlabel('Tempo (s)');
ylabel('Energia (J)');
title('Energia Dissipada no Resistor');
grid on;

% Conservação de energia
if abs(v0) < 1e-6 && abs(iL0) < 1e-6
    % Energia fornecida pela fonte: ∫I*v dt
    W_fonte = cumtrapz(t_num, I * v_num);
    W_total_sistema = W_total + W_R;
    
    subplot(2, 2, [3, 4]);
    plot(t_num, W_fonte, 'k-', 'LineWidth', 2);
    hold on;
    plot(t_num, W_total_sistema, 'r--', 'LineWidth', 2);
    xlabel('Tempo (s)');
    ylabel('Energia (J)');
    title('Conservação de Energia');
    legend('Energia fornecida pela fonte', 'Energia no sistema', 'Location', 'best');
    grid on;
    
    erro_energia = max(abs(W_fonte - W_total_sistema));
    fprintf('Erro na conservação de energia: %.2e J\n', erro_energia);
end

%% 8. COMPARAÇÃO SÉRIE vs PARALELO (se tiver dados)
% ===================================================
comparar = input('\nComparar com circuito série equivalente? (1=Sim, 0=Não): ');
if comparar == 1
    % Para comparação justa, usar componentes com mesmo valor
    fprintf('\n--- Circuito Série Equivalente ---\n');
    fprintf('Usando mesmos valores de R, L, C\n');
    fprintf('Tensão da fonte: V = I * R = %.2f V\n', I*R);
    
    % Chamar função série ou resolver diretamente
    V_serie = I * R;
    i0_serie = 0;  % Assumindo condições iniciais zero
    vC0_serie = 0;
    
    % Resolver série numericamente
    f_serie = @(t, y) [
        (V_serie - R*y(1) - y(2))/L;
        y(1)/C
    ];
    
    y0_serie = [i0_serie; vC0_serie];
    [t_serie, y_serie] = ode45(f_serie, tspan, y0_serie, options);
    i_serie = y_serie(:, 1);
    vC_serie = y_serie(:, 2);
    
    figure('Position', [100, 100, 1000, 600]);
    
    % Comparar tensão no capacitor (paralelo) vs corrente no indutor (série)
    % Normalizar para comparação
    v_par_norm = v_num / max(abs(v_num));
    i_ser_norm = i_serie / max(abs(i_serie));
    
    subplot(2, 2, 1);
    plot(t_num, v_par_norm, 'b-', 'LineWidth', 2);
    hold on;
    plot(t_serie, i_ser_norm, 'r--', 'LineWidth', 2);
    xlabel('Tempo (s)');
    ylabel('Resposta Normalizada');
    title('Comparação: Paralelo (v) vs Série (i)');
    legend('v(t) - Paralelo', 'i(t) - Série', 'Location', 'best');
    grid on;
    
    % Comparar energia
    W_C_par = 0.5 * C * v_num.^2;
    W_L_ser = 0.5 * L * i_serie.^2;
    
    subplot(2, 2, 2);
    plot(t_num, W_C_par/max(W_C_par), 'b-', 'LineWidth', 2);
    hold on;
    plot(t_serie, W_L_ser/max(W_L_ser), 'r--', 'LineWidth', 2);
    xlabel('Tempo (s)');
    ylabel('Energia Normalizada');
    title('Comparação de Energia Armazenada');
    legend('W_C - Paralelo', 'W_L - Série', 'Location', 'best');
    grid on;
    
    % Comparar taxa de amortecimento
    alpha_serie = R/(2*L);
    alpha_paralelo = 1/(2*R*C);
    
    subplot(2, 2, [3, 4]);
    bar([1, 2], [alpha_serie, alpha_paralelo]);
    set(gca, 'XTickLabel', {'Série', 'Paralelo'});
    ylabel('Coeficiente de Amortecimento α (Np/s)');
    title('Comparação dos Coeficientes de Amortecimento');
    grid on;
    
    fprintf('\n=== COMPARAÇÃO SÉRIE vs PARALELO ===\n');
    fprintf('α_serie = %.4f Np/s\n', alpha_serie);
    fprintf('α_paralelo = %.4f Np/s\n', alpha_paralelo);
    fprintf('Razão α_paralelo/α_serie = %.4f\n', alpha_paralelo/alpha_serie);
end

fprintf('\n=== ANÁLISE CONCLUÍDA ===\n');